browser/scripts/ui-core.js

```diff
@@ -67,7 +72,240 @@ VizorUI.prototype._init = function(e2) {	// called by .init() in ui.js
 };
 
+VizorUI.prototype.getZoom = function() {
+	return 1;
+}
+
+/**
+ * get the state of a UI (tabbed) panel. if no $domElement provided, then _found = false
+ * @param $domElement
+ * @returns {{_found: boolean, visible: boolean, collapsed: boolean, selectedTab: null, x: boolean, y: boolean}}
+ */
+VizorUI.prototype.getDomPanelState = function($domElement) {	/* @var $domElement jQuery */
+	var state = {
+		_found: false,
+		visible : true,
+		collapsed : false,
+		selectedTab : null,
+		x: false,
+		y: false,
+		w: -1,
+		h: -1
+	};
+	if ((typeof $domElement === 'object') && ($domElement.length > 0)) {
+		state._found = true;
+		state.visible = $domElement.is(':visible') && $domElement.hasClass('uiopen');
+		state.collapsed = $domElement.hasClass('collapsed');
+		var pos = $domElement.position();
+		if (pos) {
+			state.x = pos.left;
+			state.y = pos.top;
+		}
+		state.w = $domElement.width();
+		state.h = $domElement.height();
+		// get the active tab
+		var $activeLi = $domElement.find('li.active');
+		if ($activeLi.length>0) {
+			var tabName, tabLink;
+			tabName = $activeLi.data('tabname');			// data-tabname='presets' preferred
+			if (tabName) {
+				state.selectedTab = tabName;
+			} else {
+				tabLink = $activeLi.find('a');
+				state.selectedTab = tabLink.attr('href');	// #something
+			}
+		}
+		// else this isn't tabbed so selectedTab does not apply
+	}
+	return state;
+}
+
+VizorUI.prototype.getState = function() {
+	if (typeof clone !== 'function') return null;	// why are we here?
+	if (!this._initialised) return false;
+	var s = {};
+	s['visible'] = this.visible;
+	s['visibility'] = clone(this.visibility);		// util.js
+	s['viewmode'] = this.viewmode;
+	s['panelStates'] = {
+		chat:		null,
+		presets:	null,
+		assets:		null
+	};
+	var ps = s['panelStates'];					// ref
+	ps.chat = this.getDomPanelState(this.dom.chatWindow);
+	ps.presets = this.getDomPanelState(this.dom.presetsLib);
+	ps.assets = this.getDomPanelState(this.dom.assetsLib);
+	return s;
+};
+
+
+VizorUI.prototype.setState = function(state) {
+	if ((!this._initialised) || (typeof state !== 'object')) return false;
+	if (typeof clone !== 'function') return null;	// why are we here?
+	this.visible = state.visible;
+	this.visibility = clone(state.visibility);
+	this.updateVisibility();
+	this.setWorldEditorMode(state.viewmode === uiViewMode.world_editor);
+
+	var that=this;
+	var applyPanelState = function($el, state, collapse_callback) {
+		if ((typeof $el !== 'object') || ($el.length === 0)) return false;
+		// ignores visibility which is already applied
+		var collapsed = state.collapsed || false;
+		var selectedTab = state.selectedTab || false;
+		var x = state.x || 0;
+		var y = state.y || 0;
+		var w = state.w || 0;	// w currently ignored
+		var h = state.h || 0;	// h only applied to chat window
+		// collapse if needed
+		if (typeof collapse_callback === 'function') {	// collapse callback will collapse the window or not
```

Comment:
it's entirely self-contained, and it only deals with its own components. 
I'd make the code path entirely linear if DOM wasn't so asynchronous. I'm also considering the rest of UI implementation as it stands..